### DISCUSSION TITLE
Send MIDI clock directly from Playtime when Reaper transport isn't running

### DISCUSSION THREAD

**Discussion by TubularCorpration** [helgoboss/helgobox#1084] [Ideas]
I plan to use Playtime with hardware that has its own internal sequencers, arpeggiators and other things like LFOs that I need to send clock to continuously. As far as I can tell, the only way to send clock right now is to start Reaper's transport with Playtime (using the Play Arrangement button), with "stop/repeat playback at end of project" turned off in Reaper's playback preferences. This does work but it's kind of clunky and stops being practical if there are any items on the timeline.

It would be very useful to have an option in the column playback settings that would let you send clock directly from Playtime to any MIDI outputs assigned to that column, with a button in the UI to toggle clock globally for all columns where it was enabled and/or an option in Global Playback Settings to disable internal clock when Reaper's transport is running.

Being able to sync hardware to Playtime is absolutely critical for me to be able to use it as anything more than a sketchpad, and the workaround using Reaper's transport is better than nothing but will be a problem with more complex projects that use Reaper's timeline and Playtime at the same time.

Unless I'm missing something and there's already a way to send clock directly from Playtime itself with Reaper's transport stopped.

**Comment by helgoboss** [comment]
That gets my vote. Wanted to add that anyway. It also would enable the Launchpads and APCs to blink the LEDS in sync with the tempo.

**Comment by seantypedthis** [comment]
I have been waiting since the last alpha to request/mention something like this or location CC for the metronome position. Would love to see the beat blink, as well as my LED meter show me where I am in a measure, figured it would be about as clean to implement as all the other MIDI feedback you already have worked in.

Wanted to make sure this wasn't already in the current feature request before diving back in!

**Reply by helgoboss** [reply]
Are you sure that sending MIDI clock would achieve what you want? Doesn't sound so from here, unless your LED meters respond to MIDI clock, which would be unusual. Sounds more like you want feedback from the "Playtime: Slot seek" target ... this can show the current clip progress on an LED meter, LED ring or motor fader or text screen.

**Reply by seantypedthis** [reply]
That is one I missed!  I was actually seeing if there was a similar option for the Matrix/metronome, as opposed to an individual slot, though, so it would be constantly feeding back, even when no slot was active. Is this possible with Slot seek?

I think either would work for me, my main way of controlling sync between applications allows me to get/fabricate the "position in bar" from MIDI clock, and I can convert that 4 count to MIDI CC on a scale of 0-127 pretty easy. Just figured it'd be easier than requesting a MIDI clock generator AND CC feedback for the main Matrix transport.

**Reply by helgoboss** [reply]
I don't think that MIDI clock would help you in this case. It would send 24 ticks per quarter note. It wouldn't be a target but a source, so there wouldn't be a good way to convert it to CC (at a maximum via a JSFX put below ReaLearn when output is "FX output").

**Reply by helgoboss** [reply]
Sounds like you want this FR: https://github.com/helgoboss/helgobox/discussions/984

**Reply by seantypedthis** [reply]
That is most likely what I want, haha, might have taken me a while to find, but upvoted! I think some sort of dummy slot that was always running the "Slot seek" action would work just as well.

And I get what you're saying, that's exactly how I would use MIDI clock from Playtime; I just want a continuous clock generated that follows Playtime's transport and that's not dependent on REAPERS transport, so the signal won't stop when REAPER's transport isn't active.
I send MIDI clock out to an external application that has a patch where I can set an arbitrary time signature, usually 4/4, and will count each tick and then calculate the BPM and also give me the position in a 4/4 measure as a float32 value, which I can convert and scale to MIDI CC between 0-127, and then feed back to my controller or REAPER with a virtual MIDI cable.

It's a bit fiddly but ultimately the best way for me to deal with OSC, DMX, Link, MIDI, and any other data I might need to handle, in a reliable fashion for live compositing and visual purposes.
And really, I'm just so used to manipulating data/protocols as I've had to come up with many different solutions for REAPER refusing to _accept_ MIDI clock, only _sending_ it as a clock source. I eventually just gave up and made my own solution, sending any clock information through another applications back to REAPER as OSC data, MIDI CC, and the like.

**Reply by helgoboss** [reply]
Ah okay. You know what you are doing :) 

Did you know that ReaLearn accepts MIDI clock? There's already a MIDI clock source actually. And it's possible to map it to "Tempo". It's super experimental and probably not tight enough, but I suppose it can be improved.

**Reply by seantypedthis** [reply]
Haha, I only wish I knew what I was doing. I just appreciate there are people like you who do that I can follow! =)

And I did indeed know that ReaLearn handles MIDI Clock, an instance of ReaLearn has sat in my monitoring FX chain for a few years now handling all I can throw at it bpm/clock wise.

I think I was using MIDI clock to project tempo for sometime, CC or Note on to tap tempo action before that, but switched to ReaLearn handling OSC a little while ago to cut down on jitter from certain applications, it really has been the smoothest of all  (not to mention your OSC implementation is way more flexible than REAPERs stock implementation).

Didn't mean to hijack the thread, but thanks for this work, as always. It really does open up all of the possibilities I can think  of, and I very much appreciate it!

**Comment by hoprocker** [comment]
I'd like to second (third?) this idea. Imho, loop-based composition, at least for outboard gear, is heavily facilitated by MIDI transport and clock signals. I greatly appreciate @TubularCorpration's suggested workaround; I can now record time-synced snippets from my gear using the (Reaper) transport controls in Playtime's upper right, combined with record-arming the cell I want to record into. Sending midi sync & (optionally) transport coordinated at the cell level seems like a natural fit for Playtime.
